File: engines/google.py
import utils as utils
import requests
from bs4 import BeautifulSoup
import config as config

def google_search(query, lang = "en"):
    if(type(lang) == "NoneType"):
        lang = "en"
    url_encode_query = utils.encode_url(query)
    url = "https://www.google.com/search?q=" + url_encode_query + "&lr=lang_" + lang + "&hl=" + lang
    print(url)
    headers = config.headers
    with requests.Session() as s:
        s.post(url, headers=headers)
        response = s.get(url, headers=headers)
    soup = BeautifulSoup(response.text, 'html.parser')

    linksContainer = soup.find("div", {"id": "search"})
    linksContainers = linksContainer.find_all("div", {"class": "g"})
    links = []
    titles = []
    for a in linksContainers:
        linkTag = a.find("a", href=True)
        url = linkTag.get("href")
        titleTag = linkTag.find("h3")
        if titleTag:
            title = titleTag.getText()
        else:
            title = "No title"
        links.append(url)
        titles.append(title)



    descriptionsParents = soup.find_all("div", {"data-sncf": "1"})
    descriptions = []
    for desc in descriptionsParents:
        try:
            descriptions.append(desc.find("div").find("span").text)
        except:
            descriptions.append("No description")

    resultsDict = []
    for i in range(0, len(descriptions)):
        try:
            if links[i].startswith("http"):               
                    resultsDict.append({"title": titles[i], "description": descriptions[i], "link": links[i]})
        except Exception as e:
            print("Error: " + str(e))
    
    return resultsDict

File: engines/google.go
package engines

import (
	"github.com/PuerkitoBio/goquery"
	"example.com/Cosmos-Search/web"
)

func Search(query string, page int) ([]map[string]string, error) {
	if page == nil {
		page := 0
	}
	encodedQuery := web.UrlEncode(query)
	var url string = fmt.Sprintf("https://google.com/search?q=%s&start=%i", encodedQuery, page*10)
	var results = []map[string]string{}

	html, err := web.GetHtml(url)
	if err != nil {
		return results, err
	}
	doc, err := goquery.NewDocumentFromReader(html)
	if err != nil {
		return results, err
	}

	doc.Find(".kvH3mc").Each(func(i int, s *goquery.Selection) {
			title := s.Find("h3").Text()
			link, _ := s.Find("a").Attr("href")
			description := s.Find(".VwiC3b").Children().Last().Text()

			result := map[string]string{
				"title":       title,
				"link":        link,
				"description": description,
			}
			results = append(results, result)
		})

	return results, nil
}

